Filesystem redesign and performance improvements

JosJuice opened the pull request:
Scanning the game list without a cache can take a while when there are many games. As evidenced by games with large file systems taking significantly longer than others, a big part of this time is spent dealing with file systems. To improve the speed of scanning the game list, I made this PR. Initializing a file system now only involves three tiny disc reads and one big disc read to a `vector<u8>`, as opposed to the old code which made lots of tiny reads (many of them even going backwards!) and then constructed a bunch of objects from the values it obtained. I have also implemented better search algorithms for the two variants of `FindFileInfo`. With these changes, building the game list cache now takes about the same time for each game – the only real speed limit that's left is the limitations of the hard drive. ~~I had hoped that these changes also would make opening the game properties fast, but it can still take a second or so with large file systems because of the time it takes for the GUI to create all the tree items.~~ (EDIT: After updating to VS2015, this seems to be reasonably fast even in master.)

Another goal with this PR was to avoid implementation details of file systems being exposed to code that uses them. In the past, getting all file infos in a directory involved looping through the file system's vector of file infos and knowing that getting the size of a directory in a GC/Wii disc file system happens to produce the index of the first file info that comes after the directory. This is now taken care of inside the file system code, and code that uses file systems simply needs to use a range-based for loop on a directory to get its children. This should make it easier to both write new code that interacts with file systems and to implement new types of file systems that can be used in the same way as GC/Wii disc file systems. However, working with iterators and unique_ptrs and whatnot is still a bit new to me, so please comment if any parts of the design are bad.

While working on the changes needed for these two goals, I also tried to do as much error checking as possible. Invalid file systems should never make Dolphin crash or hang, because this code gets run when Dolphin boots. ~~The only remaining issue that I know of is that Dolphin can run out of memory if a file system is too large, which is handled by PR #2610.~~ (EDIT: That fix has been merged.)

Now, the question is... @JMC47, how much faster can your game list be scanned with this PR?

EDIT: This description was originally written when using VS2013. Now that Dolphin uses VS2015, the speed of scanning the game list and opening game properties windows has been significantly improved in master, at least with local drives. That means that the speed difference between master and this PR isn't as large as on VS2013, but it still does improve the speed, especially in situations like FileMonitor being called frequently. I'm not sure what the situation is like on other compilers/platforms.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/2820)

<!-- Reviewable:end -->

delroth commented:
@shuffle2 I seem to remember you use Dolphin with games on a networked file system. If I'm right, care to test this to see how much faster gamelist loading is?

delroth commented on Source/Core/DiscIO/FileSystemGCWii.h:
```diff
@@ -16,34 +17,81 @@ namespace DiscIO
+
+public:
+	// Set everything manually
+	CFileInfoGCWii(const u8* fst, bool wii, u32 fst_size, u32 index, u32 total_file_infos);
```
Please specify whether this takes ownership of "fst".

delroth commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
```diff
@@ -19,83 +21,309 @@
+
+u32 CFileInfoGCWii::FixIndex(u32 index) const
+{
+	if (GetFSTSize() / 0xC <= index)
```
What's 0xC here and in many places? Can you replace it with either the appropriate sizeof or a named constant?

JosJuice replied:
It's the size of an FST entry. I'll make a constant for it.

delroth commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
I tend to dislike APIs that return unique_ptr -- the caller might want to use a different type of ownership for the data, and choosing for them isn't nice :)

Can you instead make these return raw pointers and modify callers to initialize unique_ptrs (or whatever else)?

delroth commented on Source/Core/DiscIO/DiscScrubber.cpp:
```diff
@@ -326,6 +319,25 @@ bool ParsePartitionData(SPartition& _rPartition)
+	for (const DiscIO::IFileInfo& file_info : directory)
+	{
+		const std::string path = file_info.GetPath();
+		DEBUG_LOG(DISCIO, "%s", path.empty() ? "/" : path.c_str());
```
I know this is a debug log and you're just porting it, but it looks 1. very spammy; 2. very non-obvious in its meaning since it will just print a path. Can you either remove it or modify the log message to make it more explicit what's being logged?

delroth commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
```diff
@@ -19,83 +21,309 @@
+
+// For the root object only
+CFileInfoGCWii::CFileInfoGCWii(const u8* fst, bool wii, u32 fst_size)
+	: m_fst(fst),
```
This screams "delegated constructors!" to me. Could you use them to reduce duplication in your initializer lists?

JosJuice replied:
It makes the program crash, at least for this constructor. It's vital that m_fst and m_index are set to the correct values before GetSize() is called, and I guess delegated constructors evaluate all the arguments before the target constructor is called?

JosJuice replied:
I've added it to the `const CFileInfoGCWii& file_info, u32 index` constructor.

JosJuice commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
The style guide says "Try to avoid using raw pointers (pointers allocated with `new`) as much as possible", and I wouldn't want any of the callers to forget deleting the pointer. Callers wanting to use a different type of ownership is a valid point, but I don't want to return a raw pointer unless there's consensus for doing so. (cc @Lioncash?)

delroth commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
```diff
@@ -19,83 +21,309 @@
+	return m_fst + m_index * 0xC;
+}
+
+std::unique_ptr<IFileInfo> CFileInfoGCWii::Clone() const
```
Same remark here regarding unique_ptr in APIs.

delroth replied:
And as noted in the other comment thread, please ignore.

JosJuice replied:
These two methods can only be accessed by the iterator class, which does want unique_ptrs.

delroth commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
Actually there's a fair argument to be made in favor of returning unique_ptr when passing ownership: it distinguishes between passing ownership (unique_ptr, then caller can use .release() if they want to) vs. returning a pointer without ownership (in which case a raw pointer would be used).

So nevermind -- unique_ptrs in this case are fine, I was just a bit confused :)

delroth commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
Nit: on malformed FST this could crash or loop indefinitely. I would maybe add an additional exit condition & associated error logging.

delroth commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
```diff
@@ -19,83 +21,309 @@
+
+u64 CFileInfoGCWii::GetFSTSize() const
+{
+	return (u64)m_fst_size << (m_wii ? 2 : 0);
```
GetOffsetShift() ?

JosJuice replied:
Calling CFileSystemGCWii's GetOffsetShift would require a reference to that object.

delroth replied:
Is it worth redefining a GetOffsetShift() utility method in this class? It's used at least twice.

delroth commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
```diff
@@ -19,83 +21,309 @@
+	return result;
+}
+
+u64 CFileInfoGCWii::GetOffset() const { return (u64)GetRawOffset() << (m_wii ? 2 : 0); }
```
GetOffsetShift() ?

delroth commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
Maybe error log?

delroth commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
std::unique_ptr<IFileInfo> result(new ...);

Or when we have wider C++14 support (though I think we have an include in Common/ for that one): auto result = std::make_unique<IFileInfo>(new ...);

delroth commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
While at it, can you fix the arguments name here for coding style compliance? I know it's not really your code, but... :)

delroth commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
(Same here re: arguments and coding style compliance.)

delroth commented on Source/Core/DiscIO/VolumeGC.cpp:
```diff
@@ -210,31 +210,48 @@ bool CVolumeGC::LoadBannerFile() const
+	}
+
+	m_banner_file.resize(file_size);
+	if (file_system->ReadFile(file_info.get(), m_banner_file.data(), m_banner_file.size()) != file_size)
```
Would it be worth having a function to read directly into an std::vector, doing the resizing etc.?

JosJuice replied:
There is an advantage of getting the size from the file info manually here. If the file size is invalid, the function returns before ReadFile is called, meaning that we don't have to waste time and memory loading a file that potentially can be huge. Skipping the `m_banner_file.resize(file_size)` line and shortening the line after that is what such a function could simplify while not changing any behavior, and I don't think that's worth it.

JosJuice commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
The root file info is always a directory, otherwise the `CFileSystemGCWii` constructor will not set `m_Valid = true`. I suppose there is a way for a malformed root entry to have a size that is too small, though. I will amend `FixIndex` to only allow indexes that are smaller than the root entry's size.

lioncash commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
```diff
@@ -19,83 +21,309 @@
+// Each entry contains 3 u32s (0xC bytes).
+u32 CFileInfoGCWii::Get(int n) const { return Common::swap32(*((u32*)(m_fst + m_index * 0xC) + n)); }
+u32 CFileInfoGCWii::GetNameOffset() const { return Get(0); }
+u32 CFileInfoGCWii::GetRawOffset() const { return Get(1); }
```
These functions should obey the brace on newline style (same with the others)

delroth replied:
Does it really matter for one-liners like that? IMO putting this on one line is more readable.

lioncash replied:
Those are acceptable within the header

JosJuice commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
I ended up adding an exit condition here instead.

JMC47 commented:
The biggest delay now is actually after it finishes scanning; there's a 5+ second delay before the gamelist actually loads.  Everything else is lightning fast for thousands of games (roughly 3 seconds?)

shuffle2 commented:
@delroth I haven't done that for a while (I was doing that when I was doing dev work on a laptop - don't really do that anymore...)

lioncash commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
couldn't you just do 

```cpp
return FindFileInfo(rest_of_path, child);
```

considering the fail case below this returns nullptr anyway?

phire commented:
I have since upgraded to a system with games on a networked filesystem (to move my spinning drivers further away from me)

Loads a lot faster, I think it's even loading faster than when I had a spinning drive in the same computer.

phire commented:
Can we get this rebased, so it can fail on all our buildbots?

JosJuice commented:
Rebased. The merge conflicts are fixed, and I have improved the error checking.

degasus commented:
@JosJuice How did you manage to be so "nice" and to not bother everyone else to review this work which is done for three months?

Kolano commented:
:( Aw, I had gotten hopeful. Compared this with 4.0-8076. At best only a ~15sec improvement on my Wii title scan (1509 files on a NAS). 2:59 vs 2:44, and follow up runs took just as long. Still wish we offered an option to fully cache the list and only rescan when clicking "Refresh".

JosJuice commented:
@degasus Um... I don't who I can ask about it and I don't know when it's appropriate? :(

@Kolano What do you mean by "fully cache"? Do you want Dolphin to not check which files are present on the disk? If a cache is available, that's the only thing Dolphin hits the disk for. (Well, long story short, it actually also reads 8 bytes from every Wii game that didn't have save data the last time the game list was read... I might fix that in another PR sometime.) Was your performance test done with an empty cache?

Kolano commented:
@JosJuice Yes, to not recheck for files on disk unless asked to, otherwise just relist previously listed titles. I didn't take steps to clear the cache, so that's likely how long it takes to scan my Wii titles /w a cache in place (I don't have save states for any but a few of the titles). What would I need to do to clear the cache (purge my Dolphin user folder)?

JosJuice commented:
In the Tools menu, there's an item called "Purge Cache" or "Purge Game List Cache" (depending on which version of Dolphin you're using). Press it and measure how long it takes for Dolphin to rebuild the cache.

Kolano commented:
Sorry for my confusion, "Purge cache" is under the "View" menu and I don't look there much.

4.0-8076: 
	Purge Cache Trial 1: 4:18
	Refresh Trial 1: 3:10

	Purge Cache Trial 2: 4:15
	Refresh Trial 2: 3:08

4.0-8183 (pr-2820):
	Purge Cache Trial 1:  4:05
	Refresh Trial 1: 3:07

	Purge Cache Trial 2: 4:13
	Refresh Trial 2: 3:47

So only a few seconds improvement for me, which is probably within the error range of these tests. So back to loading things in a title at a time to avoid a ~4m wait at Dolphin startup.

lioncash commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[Source/Core/DiscIO/Filesystem.h, line 24 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMnVeLgxrQiHoBDlTD:-KLMnVeLgxrQiHoBDlTE:-2122588331) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/Filesystem.h#L24)):*
> ```objc
> 
> public:
>   class const_iterator final : public std::iterator<std::input_iterator_tag, IFileInfo>
> ```

This is a non-compliant input_iterator since it doesn't define the postfix `operator++`. 

An input_iterator also shouldn't typically be moving its contents (or particularly allowing it), since the intent of the basic set of iterators is to act as a singular view across a data-set. It shouldn't be directly altering the state of what it observes (or act as a carrier for allocated objects). An iterator should simply iterate, not manage lifetimes.

If any of these were passed to standard library functions taking iterators, it would churn memory like crazy due to all of the allocations and deallocations happening.

---

*[Source/Core/DiscIO/Filesystem.h, line 59 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMwZ4Z4Lt60GQEZ7Ln:-KLMwZ4Z4Lt60GQEZ7Lo:933865438) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/Filesystem.h#L59)):*
> ```objc
>   bool operator==(const IFileInfo& other) const { return GetAddress() == other.GetAddress(); }
>   bool operator!=(const IFileInfo& other) const { return !operator==(other); }
>   virtual IFileInfo& operator++() = 0;
> ```

This is very suspect, especially if it's in the public interface especially because begin() and end() already exist. For example, std::vector and std::list don't expose anything like this.

---

*[Source/Core/DiscIO/Filesystem.h, line 62 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMmzWFWLDIqw-RzVn4:-KLMmzWGgsu1Hc95geNo:-1072162998) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/Filesystem.h#L62)):*
> ```objc
>   virtual std::unique_ptr<IFileInfo> clone() const = 0;
>   virtual const_iterator begin() const = 0;
>   virtual const_iterator end() const = 0;
> ```

container implementations typically also provide `cbegin` and `cend`for const iterators (which allows non-member `std::cbegin` and `std::cend` to function with them)

---

*[Source/Core/DiscIO/Filesystem.h, line 82 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLN1eNYwS11BOj0eOoi:-KLN1eNZiUyCbA_oLp0h:419538741) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/Filesystem.h#L82)):*
> ```objc
> protected:
>   // Used for comparisons with other file info objects
>   virtual const void* GetAddress() const = 0;
> ```

This should be a concrete type, not `void*`. Throwing away type information isn't necessary. This should likely be a `u8*` or `char*`.

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 89 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMm1mSOLBBW4HBD17f:-KLMm1mTCfisKoWvZFmn:1041501815) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/FileSystemGCWii.cpp#L89)):*
> ```cpp
> u32 CFileInfoGCWii::GetRawNameOffset() const
> {
>   return Get(0);
> ```

These should be named values rather than arbitrary magic values.

---

*[Source/Core/DiscIO/FileSystemGCWii.h, line 50 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMmqNpl1BIDMyeu-i8:-KLMmqNqFjAsxb0TFHfV:-280465871) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/FileSystemGCWii.h#L50)):*
> ```objc
>   std::string GetPath() const override;
> 
> 	bool IsValid(u64 fst_size, const CFileInfoGCWii& parent_directory) const;
> ```

Random tab.

---

*[Source/Core/DiscIO/FileSystemGCWii.h, line 56 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMmiXY0u5zD2g0Hj7A:-KLMmiXY0u5zD2g0Hj7B:-525064684) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/FileSystemGCWii.h#L56)):*
> ```objc
> 
> private:
>   u32 GetNextIndex() const;
> ```

This does different things for directories and files, so this should likely be documented as well.

---

*[Source/Core/DiscIO/FileSystemGCWii.h, line 58 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMmNErWnESh_3kP342:-KLMmNEs4Zwg23ySoPhu:1430798530) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/FileSystemGCWii.h#L58)):*
> ```objc
>   u32 GetNextIndex() const;
> 
>   u32 Get(int n) const;
> ```

`n` doesn't mean anything.

---

*[Source/Core/DiscIO/VolumeWiiCrypted.cpp, line 245 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMqzYU7kVxDVT1HLIe:-KLMqzYVF82O5rld2shS:-290787204) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/VolumeWiiCrypted.cpp#L245)):*
> ```cpp
>   std::unique_ptr<IFileSystem> file_system(CreateFileSystem(this));
>   if (!file_system)
>     return std::map<IVolume::ELanguage, std::string>();
> ```

`return {};`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820)*
<!-- Sent from Reviewable.io -->

BhaaLseN commented:
Review status: 0 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 81 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLN8y7_JGmc6OFdBC-Y:-KLN8y7_JGmc6OFdBC-Z:1576159610) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/FileSystemGCWii.cpp#L81)):*
> ```cpp
> }
> 
> // Reads the first u32 of this entry for Get(0), reads the u32 after that for Get(1), and so on
> ```

"Reads the n'th (zero-based) u32 of this entry..." as additional comment perhaps?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
@dolphin-emu-bot rebuild

JosJuice commented:
Review status: 0 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[Source/Core/DiscIO/Filesystem.h, line 24 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMnVeLgxrQiHoBDlTD:-KMuW97bMKJVi6Tw8uq7:676529800) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/Filesystem.h#L24)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This is a non-compliant input_iterator since it doesn't define the postfix `operator++`. 
> 
> An input_iterator also shouldn't typically be moving its contents (or particularly allowing it), since the intent of the basic set of iterators is to act as a singular view across a data-set. It shouldn't be directly altering the state of what it observes (or act as a carrier for allocated objects). An iterator should simply iterate, not manage lifetimes.
> 
> If any of these were passed to standard library functions taking iterators, it would churn memory like crazy due to all of the allocations and deallocations happening.
> 
> 
</details>
Thanks for the feedback. I need some more time to think about this, so I will respond later.

---

*[Source/Core/DiscIO/Filesystem.h, line 82 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLN1eNYwS11BOj0eOoi:-KMuVd5rujbCSWZWxgn5:-1534825560) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/Filesystem.h#L82)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This should be a concrete type, not `void*`. Throwing away type information isn't necessary. This should likely be a `u8*` or `char*`.
</details>
The reason I didn't want to use a concrete pointer type is because the returned value only is supposed to be compared to other values, not be dereferenced. Would `uintptr_t` be appropriate for that? Or should I change to `const u8*`?

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 81 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLN8y7_JGmc6OFdBC-Y:-KMuHjvZPt0lcCFICXO9:1574904165) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/FileSystemGCWii.cpp#L81)):*
<details><summary><i>Previously, BhaaLseN (BhaaL) wrote…</i></summary>
> "Reads the n'th (zero-based) u32 of this entry..." as additional comment perhaps?
</details>
Now that an enum is used, I removed this comment. Does it need additional explanation the way it is now?

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 89 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMm1mSOLBBW4HBD17f:-KMuHhw4YX0vmoj9aSwM:-1727218310) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/FileSystemGCWii.cpp#L89)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> These should be named values rather than arbitrary magic values.
</details>
I've made Get use an enum.

---

*[Source/Core/DiscIO/FileSystemGCWii.h, line 50 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMmqNpl1BIDMyeu-i8:-KMuDt4I7U985W44R3r9:1758359711) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/FileSystemGCWii.h#L50)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Random tab.
</details>
Fixed.

---

*[Source/Core/DiscIO/FileSystemGCWii.h, line 56 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMmiXY0u5zD2g0Hj7A:-KMuVYiSWrlem2UTEpwm:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/FileSystemGCWii.h#L56)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This does different things for directories and files, so this should likely be documented as well.
</details>
Done.

---

*[Source/Core/DiscIO/FileSystemGCWii.h, line 58 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMmNErWnESh_3kP342:-KMuITFYVp2uiGOu9cpT:1564406616) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/FileSystemGCWii.h#L58)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `n` doesn't mean anything.
</details>
It has been changed.

---

*[Source/Core/DiscIO/VolumeWiiCrypted.cpp, line 245 \[r15\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMqzYU7kVxDVT1HLIe:-KMuDkxSc83pivDO1ACf:-508806595) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/VolumeWiiCrypted.cpp#L245)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `return {};`
</details>
Done, except I used {{}} instead of {} to avoid FreeBSD build failures.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[Source/Core/DiscIO/Filesystem.h, line 24 at r15](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMnVeLgxrQiHoBDlTD:-Kc9Qkp6Qk3AuDdqQwg7:b-fcxns6) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/Filesystem.h#L24)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary><blockquote>

Thanks for the feedback. I need some more time to think about this, so I will respond later.
</blockquote></details>

I was intending to reply a few days later, not a few months later... But better late than never, I guess :)

I've added a postfix `operator++`.

The intent of this iterator was to provide a view over (part of) an IFileSystem. However, there's something that complicates it: There can be different kinds of subclasses to IFileSystem. (There's only one right now, but it would be nice to add WAD support in the future.) Since the iterator type has to be returned by value, I can't make subclasses of the iterator to match the different subtypes of IFileSystem. It works out when I have IFileInfo and its subtypes "in between", though. Though I guess that even if we only wanted to support one kind of file system, the current design would still require IFileInfo since we don't want users of IFileSystem to hit the vector<u8> directly... But I suppose we at least could have IFileInfo directory inside iterators then instead of unique_ptr<IFileInfo>. I dunno. Do you think there's a better way to let people iterate through different types of filesystems using a single iterator type? Or am I trying to do this a completely wrong way? I agree that it's a bet messy right now (especially with the extra memory allocations), but iterators make iterating through directories so nice, and I don't know how iterators should be implemented for this otherwise...

---

*[Source/Core/DiscIO/Filesystem.h, line 59 at r15](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMwZ4Z4Lt60GQEZ7Ln:-Kc9UgBWtsxiZpS1lm3F:bdm32d4) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/Filesystem.h#L59)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary><blockquote>

This is very suspect, especially if it's in the public interface especially because begin() and end() already exist. For example, std::vector and std::list don't expose anything like this.
</blockquote></details>

I've made it protected.

---

*[Source/Core/DiscIO/Filesystem.h, line 62 at r15](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLMmzWFWLDIqw-RzVn4:-Kc9TSosDqBcRWXKRItA:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/Filesystem.h#L62)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary><blockquote>

container implementations typically also provide `cbegin` and `cend`for const iterators (which allows non-member `std::cbegin` and `std::cend` to function with them)
</blockquote></details>

Done.

---

*[Source/Core/DiscIO/Filesystem.h, line 82 at r15](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-KLN1eNYwS11BOj0eOoi:-Kc9TTEzBtR09eQww-CZ:b3kxmgl) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e4309d441f19f686a289ee75fdc1df09471092c0/Source/Core/DiscIO/Filesystem.h#L82)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary><blockquote>

The reason I didn't want to use a concrete pointer type is because the returned value only is supposed to be compared to other values, not be dereferenced. Would `uintptr_t` be appropriate for that? Or should I change to `const u8*`?
</blockquote></details>

I made it use `uintptr_t`. Please tell me if it's inappropriate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 0 of 12 files reviewed at latest revision, 10 unresolved discussions.

---

*[Source/Core/DiscIO/Filesystem.h, line 27 at r19](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Kc9pfxfjM4dOTqYVQqH:-Kc9pfxfjM4dOTqYVQqI:b9km5l8) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e7d9e9e8d6e3a7d430bdb602e48b805e6073ce38/Source/Core/DiscIO/Filesystem.h#L27)):*
> ```objc
>   {
>   public:
>     using iterator_category = std::input_iterator_tag;
> ```

Should I make this a forward iterator? The only thing that's missing as far as I know is a default constructor, which isn't especially hard to write (just set m_file_info to nullptr), but then I would need to add nullptr checks to functions like `operator==`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820)*
<!-- Sent from Reviewable.io -->

leoetlino commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
```diff
@@ -20,11 +20,39 @@
+
+u64 CFileInfoGCWii::GetOffset() const
+{
+  return (u64)Get(EntryProperty::FILE_OFFSET) << (m_wii && !IsDirectory() ? 2 : 0);
```
`static_cast<u64>`

sepalani commented on Source/Core/DiscIO/VolumeGC.cpp:
```diff
@@ -193,8 +193,14 @@ void CVolumeGC::LoadBannerFile() const
+  if (!file_info)
+    return;
 
+  size_t file_size = (size_t)file_info->GetSize();
```
static_cast.

JosJuice replied:
This was necessary in the past because GetSize used to return u64.

In practice, I guess we won't run into any system where `sizeof(size_t) < sizeof(u32)`, but the C++ standard doesn't guarantee `sizeof(size_t) >= sizeof(u32)` as far as I know. @lioncash, should the cast be removed or not? (If not, I'll make it a `static_cast`.)

lioncash replied:
`size_t` is guaranteed to be able to safely store the size of the largest possible theoretical object allowed on a system (exception: it cannot be used to store addresses to member-function pointers). On a 32-bit system this will be a 32-bit value, on 64-bit it'll be a 64-bit value (as the type needs to be able to cover the whole memory range).

JosJuice replied:
Yes, but are we assuming that all systems that run Dolphin will be at least 32-bit? (I would assume so since the amount of RAM we use when emulating requires about 30 bits to be addressed, but I thought I should check to make sure)

lioncash replied:
I'd probably just leave it in for the generic target or something.

sepalani commented on Source/Core/DiscIO/FileSystemGCWii.cpp:
```diff
@@ -20,131 +22,393 @@
   {
     // Limit read size to 128 MB
-    size_t readSize = (size_t)std::min(remainingSize, (u64)0x08000000);
+    size_t read_size = (size_t)std::min(remaining_size, (u64)0x08000000);
```
static_cast.

JosJuice replied:
The only reason I'm touching this part of the code is to change the variable naming style throughout all of the filesystem code (as suggested by delroth). I'd prefer to not need to make any additional changes just because of the renaming commit - this PR is already big enough as it is.

sepalani commented on Source/Core/DolphinWX/ISOProperties/FilesystemPanel.cpp:
```diff
@@ -23,6 +25,8 @@
 #include "Common/Logging/Log.h"
 #include "DiscIO/Enums.h"
 #include "DiscIO/Filesystem.h"
+// TODO: eww
```
It needs a bit more explicit details, maybe.

JosJuice replied:
This (including the include right below) was supposed to be removed in one of the later commits. The removal must've gotten lost in the ISOProperties-to-FilesystemPanel rebase. I'll remove it later today.

lioncash commented:
Replying to the following comment after 8 months.

>I was intending to reply a few days later, not a few months later... But better late than never, I guess :)
>
>I've added a postfix operator++.
>
>The intent of this iterator was to provide a view over (part of) an IFileSystem. However, there's something that complicates it: There can be different kinds of subclasses to IFileSystem. (There's only one right now, but it would be nice to add WAD support in the future.) Since the iterator type has to be returned by value, I can't make subclasses of the iterator to match the different subtypes of IFileSystem. It works out when I have IFileInfo and its subtypes "in between", though. Though I guess that even if we only wanted to support one kind of file system, the current design would still require IFileInfo since we don't want users of IFileSystem to hit the vector directly... But I suppose we at least could have IFileInfo directory inside iterators then instead of unique_ptr. I dunno. Do you think there's a better way to let people iterate through different types of filesystems using a single iterator type? Or am I trying to do this a completely wrong way? I agree that it's a bet messy right now (especially with the extra memory allocations), but iterators make iterating through directories so nice, and I don't know how iterators should be implemented for this otherwise...

I don't particularly fully understand the code layout as it currently is (in terms of implementation), so I find it hard to comment on this in an adequate way. Have you tried looking at libstdc++ or libc++'s implementation of the iterators for std::filesystem to get a different viewpoint?

JosJuice commented:
I hadn't looked at them before, but I took a look now. In both of them, `directory_iterator` contains a `shared_ptr` to an internal class (`_Dir` for libstdc++, `__dir_stream` for libc++). The implementation of `directory_iterator::operator++` calls the `advance` function in the internal class, which is what actually calls the OS's API (`readdir` or `FindNextFile`). So when it comes to the design question we were talking about, the major differences between those implementations and my implementation is that I'm using `unique_ptr` instead of `shared_ptr`, that my equivalent of `advance` is virtual (because I want to allow supporting more than one type of filesystem at once) (but I'm guessing the OS might do something similar inside the call to readdir/FindNextFile anyway), and that I have `CFileInfoGCWii` hit the raw bytes of the FST `vector` instead of calling OS functions (since there obviously aren't any OS functions for this).

Would switching from `unique_ptr` to `shared_ptr` be a good choice? (I'm not sure how often these kinds of iterators typically will be copied.)

leoetlino commented:
I'm not 100% sure about the iterator usage either, but it looks similar to the std::filesystem implementation of iterators.

Other changes mostly look good, except a few things.

---

Reviewed 1 of 9 files at r20, 9 of 13 files at r22, 4 of 4 files at r23.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 27 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1uIZBl4BZbY8EWp_J:-Km1uIZBl4BZbY8EWp_K:b-ypv55p) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.cpp#L27)):*
> ```cpp
> namespace DiscIO
> {
> static constexpr u32 FST_ENTRY_SIZE = 4 * 3;  // An FST entry consists of three 32-bit integers
> ```

There is no need for static here.

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 48 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1uOZfSKJimeR5o1Ag:-Km1uOZfSKJimeR5o1Ah:bxd39x5) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.cpp#L48)):*
> ```cpp
> }
> 
> FileInfoGCWii::~FileInfoGCWii()
> ```

= default

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 255 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1vRQV11w7EbJT96KF:-Km1vRQV11w7EbJT96KG:bxd39x5) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.cpp#L255)):*
> ```cpp
> }
> 
> FileSystemGCWii::~FileSystemGCWii()
> ```

= default

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 351 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1pt_TALd79pNAiJFO:-Km1pt_TALd79pNAiJFP:b-4va2hm) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.cpp#L351)):*
> ```cpp
>                     " Size: %" PRIx64,
>             read_length, offset_in_file, file_info->GetPath().c_str(), file_info->GetOffset(),
>             file_info->GetSize());
> ```

GetSize() returns a u32, not a u64.

---

*[Source/Core/DiscIO/FileSystemGCWii.h, line 96 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1y9pZLDhm3z_xV7eD:-Km1y9pZLDhm3z_xV7eE:bo3mv7c) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.h#L96)):*
> ```objc
>   std::unique_ptr<FileInfo> FindFileInfo(u64 disc_offset) const override;
> 
>   u64 ReadFile(const FileInfo* file_info, u8* buffer, u64 max_buffer_size,
> ```

Is there a reason for using a pointer here instead of a const ref? 

---

*[Source/Core/DolphinWX/ISOProperties/FilesystemPanel.cpp, line 434 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1q3Ifl8YY8NcsnTkb:-Km1q3Ifl8YY8NcsnTkc:b6s4u3t) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DolphinWX/ISOProperties/FilesystemPanel.cpp#L434)):*
> ```cpp
>       return;
> 
>     DEBUG_LOG(DISCIO, output_path.c_str());
> ```

This is potentially unsafe. You should use "%s" for the format string.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-:-Km1zebQu4uRWCnS8xKV:brgunvn)*
<!-- Sent from Reviewable.io -->

leoetlino commented:
Btw, it looks like the filesystem panel is broken. No entry is shown at all (both for GC and Wii games).

leoetlino commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 201 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km23HcxQ5T92BcZnOcu:-Km23HcxQ5T92BcZnOcv:b-tsslix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.cpp#L201)):*
> ```cpp
>   const std::optional<u32> fst_offset_unshifted = m_volume->ReadSwapped<u32>(0x424, m_partition);
>   const std::optional<u32> fst_size_unshifted = m_volume->ReadSwapped<u32>(0x428, m_partition);
>   if (!fst_offset_unshifted || fst_size_unshifted)
> ```

Bug -- this should be: `if (!fst_offset_unshifted || !fst_size_unshifted)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 27 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1uIZBl4BZbY8EWp_J:-KmY-rJmHZ-Cu1nAeOn5:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.cpp#L27)):*
<details><summary><i>Previously, leoetlino (Leo Lam) wrote…</i></summary><blockquote>

There is no need for static here.
</blockquote></details>

Done.

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 48 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1uOZfSKJimeR5o1Ag:-KmY18odlqw-gMogFHBB:b-bdwg4q) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.cpp#L48)):*
<details><summary><i>Previously, leoetlino (Leo Lam) wrote…</i></summary><blockquote>

= default
</blockquote></details>

Should I change the Filesystem.h destructors too to be consistent? And should the `= default`s be in the .h file or .cpp file?

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 201 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km23HcxQ5T92BcZnOcu:-KmXy-wkQhjsw4m6izOm:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.cpp#L201)):*
<details><summary><i>Previously, leoetlino (Leo Lam) wrote…</i></summary><blockquote>

Bug -- this should be: `if (!fst_offset_unshifted || !fst_size_unshifted)`
</blockquote></details>

Done.

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 351 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1pt_TALd79pNAiJFO:-KmXy47Dp7nGhn6tCatN:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.cpp#L351)):*
<details><summary><i>Previously, leoetlino (Leo Lam) wrote…</i></summary><blockquote>

GetSize() returns a u32, not a u64.
</blockquote></details>

Done.

---

*[Source/Core/DiscIO/FileSystemGCWii.h, line 96 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1y9pZLDhm3z_xV7eD:-KmXAjnnvOYMcPozkEEw:b-lhp7zk) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.h#L96)):*
<details><summary><i>Previously, leoetlino (Leo Lam) wrote…</i></summary><blockquote>

Is there a reason for using a pointer here instead of a const ref? 
</blockquote></details>

It can let callers skip checking whether the `FileInfo` returned by `FindFileInfo` is valid. See VolumeWii.cpp for an example.

---

*[Source/Core/DolphinWX/ISOProperties/FilesystemPanel.cpp, line 434 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1q3Ifl8YY8NcsnTkb:-KmXAHc2SxmS1GEhkxY4:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DolphinWX/ISOProperties/FilesystemPanel.cpp#L434)):*
<details><summary><i>Previously, leoetlino (Leo Lam) wrote…</i></summary><blockquote>

This is potentially unsafe. You should use "%s" for the format string.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820)*
<!-- Sent from Reviewable.io -->

leoetlino commented:
Reviewed 1 of 6 files at r24, 2 of 4 files at r25.
Review status: 9 of 12 files reviewed at latest revision, 11 unresolved discussions.

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 48 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1uOZfSKJimeR5o1Ag:-KmY5o9q_zcnE95HGAfS:bjatodh) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.cpp#L48)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary><blockquote>

Should I change the Filesystem.h destructors too to be consistent? And should the `= default`s be in the .h file or .cpp file?
</blockquote></details>

Yes, please change the other destructors too. I think it'd be better for them to be in the .cpp file, so if you need to change the destructor, you don't need to recompile everything using the header.

---

*[Source/Core/DiscIO/FileSystemGCWii.h, line 96 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1y9pZLDhm3z_xV7eD:-KmY6SkpWz1S7QS-80V_:b-1xf9ie) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.h#L96)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary><blockquote>

It can let callers skip checking whether the `FileInfo` returned by `FindFileInfo` is valid. See VolumeWii.cpp for an example.
</blockquote></details>

Ah, okay.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 9 of 12 files reviewed at latest revision, 10 unresolved discussions.

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 48 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1uOZfSKJimeR5o1Ag:-KmYGg33b6E1_xnLs26Y:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.cpp#L48)):*
<details><summary><i>Previously, leoetlino (Leo Lam) wrote…</i></summary><blockquote>

Yes, please change the other destructors too. I think it'd be better for them to be in the .cpp file, so if you need to change the destructor, you don't need to recompile everything using the header.
</blockquote></details>

Done.

---

*[Source/Core/DiscIO/FileSystemGCWii.cpp, line 255 at r23](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Km1vRQV11w7EbJT96KF:-KmYGj3-I3NSP1_jtPy0:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/919db31cfa675dd6c74b1f03eac60977a60c871c/Source/Core/DiscIO/FileSystemGCWii.cpp#L255)):*
<details><summary><i>Previously, leoetlino (Leo Lam) wrote…</i></summary><blockquote>

= default
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820)*
<!-- Sent from Reviewable.io -->

leoetlino commented:
Reviewed 2 of 3 files at r26.
Review status: 10 of 12 files reviewed at latest revision, 10 unresolved discussions.

---

*[Source/Core/DiscIO/Filesystem.h, line 27 at r19](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Kc9pfxfjM4dOTqYVQqH:-Km_clp75nzSU0aBvI-S:bkjse35) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e7d9e9e8d6e3a7d430bdb602e48b805e6073ce38/Source/Core/DiscIO/Filesystem.h#L27)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary><blockquote>

Should I make this a forward iterator? The only thing that's missing as far as I know is a default constructor, which isn't especially hard to write (just set m_file_info to nullptr), but then I would need to add nullptr checks to functions like `operator==`.
</blockquote></details>

Looks like I had missed this comment. I don't see why not, since this iterator satisfies nearly all requirements (including multipass) and you really just have to make it default constructible (+ the checks you mentioned).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 10 of 12 files reviewed at latest revision, 10 unresolved discussions.

---

*[Source/Core/DiscIO/Filesystem.h, line 27 at r19](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820#-Kc9pfxfjM4dOTqYVQqH:-Kmam8QJ0QtLSwquEkAz:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e7d9e9e8d6e3a7d430bdb602e48b805e6073ce38/Source/Core/DiscIO/Filesystem.h#L27)):*
<details><summary><i>Previously, leoetlino (Leo Lam) wrote…</i></summary><blockquote>

Looks like I had missed this comment. I don't see why not, since this iterator satisfies nearly all requirements (including multipass) and you really just have to make it default constructible (+ the checks you mentioned).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820)*
<!-- Sent from Reviewable.io -->

leoetlino commented:
Reviewed 1 of 6 files at r24, 1 of 3 files at r26, 1 of 1 files at r27.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/2820)*
<!-- Sent from Reviewable.io -->
